core/src/main/java/com/linecorp/armeria/common/stream/StreamMessageInputStream.java

```diff
@@ -0,0 +1,149 @@
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common.stream;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.function.Function;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.internal.common.stream.StreamMessageUtil;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+
+final class StreamMessageInputStream<T> extends InputStream {
+
+    public static <T> StreamMessageInputStream<T> of(StreamMessage<T> source,
+                                                     Function<T, Byte> byteConverter) {
+        return new StreamMessageInputStream<>(source, byteConverter);
+    }
+
+    private final ByteBuf buf = ByteBufAllocator.DEFAULT.buffer();
+    @Nullable
+    private final StreamMessageInputStreamSubscriber<T> subscriber;
+
+    private StreamMessageInputStream(StreamMessage<T> source,
+                                     Function<T, Byte> byteConverter) {
+        requireNonNull(source, "source");
+        requireNonNull(byteConverter, "byteConverter");
+        subscriber = new StreamMessageInputStreamSubscriber<>(buf, byteConverter);
+        source.subscribe(subscriber);
+    }
+
+    @Override
+    public int read() throws IOException {
+        return buf.isReadable() ? (buf.readByte() & 0xFF) : -1;
+    }
+
+    @Override
+    public int read(byte[] dst, int dstIndex, int len) throws IOException {
+        if (!buf.isReadable()) {
+            return -1;
+        }
+        if (len <= 0) {
+            return 0;
+        }
+        len = Math.min(len, buf.readableBytes());
+        buf.readBytes(dst, dstIndex, len);
+        return len;
+    }
+
+    @Override
+    public void close() throws IOException {
+        subscriber.cancel();
+        buf.clear();
+    }
+
+    @Override
+    public int available() {
+        return buf.readableBytes();
+    }
+
+    private static final class StreamMessageInputStreamSubscriber<T> implements Subscriber<T>, Subscription {
+
+        private final ByteBuf buf;
+        private final Function<T, Byte> byteConverter;
+        @Nullable
+        private volatile Subscription upstream;
+        private volatile boolean canceled;
+
+        StreamMessageInputStreamSubscriber(ByteBuf buf, Function<T, Byte> byteConverter) {
+            requireNonNull(buf, "buf");
+            requireNonNull(byteConverter, "byteConverter");
+            this.buf = buf;
+            this.byteConverter = byteConverter;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            upstream = subscription;
+            upstream.request(Long.MAX_VALUE);
+        }
```

Comment:
```java
// Q. where to `upstream.request(N)`?

// option 1. onSubscribe 
public void onSubscribe() {
  upstream.request(Long.MAX_VALUE);
}

// option 2. onNext
public void onNext() {
  upstream.request(1);
}

```
(curiosity) is it better to `upstream.request(1)` on `onNext` instead of `onSubscribe`? 🤔

Reply:
We have to request an element one by one only when there is demand for data.